File: 0155-min-stack/0155-min-stack.py
class MinStack:
    def __init__(self):
        self.stack = list()
        self.minimum = list()

    def push(self, val: int) -> None:
        if not len(self.minimum): self.minimum.append(val)
        else: self.minimum.append(min(self.minimum[-1], val))
        self.stack.append(val)

    def pop(self) -> None:
        self.minimum.pop()
        return self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minimum[-1]



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    stack<int> stk;
    stack<int> minimum;

    MinStack() {
        return;
    }
    
    void push(int val) {
        if (minimum.empty()) {
            minimum.push(val);
        } else {
            minimum.push(min(val, minimum.top()));
        }
        stk.push(val);
    }
    
    void pop() {
        minimum.pop();
        stk.pop();
    }
    
    int top() {
        return stk.top();
    }
    
    int getMin() {
        return minimum.top();
    }
};
